src/core/server/saved_objects/validation/validator.ts

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { ValidationError, ObjectType, isConfigSchema } from '@kbn/config-schema';
+import { SavedObjectAttributes } from '../../types';
+import { SavedObjectsValidationError } from './validator_error';
+
+/**
+ * The custom validation function if @kbn/config-schema is not a valid solution for your specific plugin requirements.
+ *
+ * @example
+ * The validation should look something like:
+ * ```typescript
+ * const myAttributesValidation: SavedObjectsValidationFunction = (data) => {
+ *   if (typeof data.bar !== 'string') {
+ *     throw new Error(`[bar]: expected value of type [string] but got [${typeof data.bar}]`);
+ *   }
+ * }
+ * ```
+ *
+ * @public
+ */
+export type SavedObjectsValidationFunction<
+  A extends SavedObjectAttributes = SavedObjectAttributes
+> = (data: A) => void;
+
+/**
+ * Allowed property validation options: either @kbn/config-schema validations or custom validation functions.
+ *
+ * See {@link SavedObjectsValidationFunction} for custom validation.
+ *
+ * @public
+ */
+export type SavedObjectsValidationSpec<A extends SavedObjectAttributes> =
+  | ObjectType
+  | SavedObjectsValidationFunction<A>;
+
+/**
+ * A map of {@link SavedObjectsValidationSpec | validation specs} to be used for a given type.
+ * The map's keys must be valid semver versions.
+ *
+ * Any time you change the schema of a {@link SavedObjectsType}, you should add a new entry
+ * to this map for the Kibana version the change was introduced in.
+ *
+ * @example
+ * ```typescript
+ * const validationMap: SavedObjectValidationMap = {
+ *   '1.0.0': schema.object({
+ *     foo: schema.string(),
+ *   }),
+ *   '1.1.0': schema.object({
+ *     foo: schema.oneOf([schema.string(), schema.boolean()]),
+ *   }),
+ *   '2.1.0': (data) => {
+ *     if (typeof data.bar !== 'string') {
+ *       throw new Error(`[bar]: expected value of type [string] but got [${typeof data.bar}]`);
+ *     }
+ *     if (typeof data.foo !== 'string' && typeof data.foo !== 'boolean') {
+ *       throw new Error(`[foo]: expected value of type [string,boolean] but got [${typeof data.foo}]`);
+ *     }
+ *   }
```

Comment:
I thought it would be a good idea to allow for custom validation functions, similar to what we do for http routes, because it feels like an inevitability that someone would eventually ask to do something that can't be handled by `@kbn/config-schema`.

However, implementing the custom functions just raised more questions that I'd like to get others' thoughts on:
1. IMO a validation function isn't the right place for folks to actually _change_ the contents of the objects, so for now I structured the validation functions differently from how they are set up for http routes... I didn't include a factory for returning `ok(value)` or `badRequest()`. It's just a function that exists to throw errors, and otherwise returns nothing. It doesn't let you modify or do anything with the data.
    - Is it worth trying to follow the same factory pattern for consistency and keeping safeguards to prevent mutation, or do we prefer this simpler approach?
2. Do we think users need access to anything other than the attributes? Currently the SO repository just passes attributes in. But there may be situations for custom functions where it is useful to have more info about the object. 
    - A more future-proof alternative might be providing `({ attributes })` to the validation functions, so that we can add more later if it proves to be necessary, WDYT?

Reply:
I don't have an opinion on (1) yet.

(2) I think we should rather pass in the entire saved object, would there be any downside? Might not be a strong use case, but theoretically you could want to do validation on the references e.g. a dashboard can't have references to a `config` saved object type or on the `updated_at` field, e.g. can't set a date in the future.

Reply:
> Is it worth trying to follow the same factory pattern for consistency and keeping safeguards to prevent mutation, or do we prefer this simpler approach

Ihmo let's KISS for now, and just document that custom validation function should *not* be used to mutate the data. If at some point we discover that API consumers are doing, or planning to do, so, we can then put real safeguards.

> (2) I think we should rather pass in the entire saved object, would there be any downside?

I agree that it would probably be helpful to allow owners to validate more than the attributes, **BUT** this means that custom validation functions would be more powerful than schema validation, which is an idea I hate to be honest, as ideally config-schema validation should always be favored unless there is a very specific reason to use custom validation instead. This is what we did for route validation for instance.

So if we think that it make sense to validate more than the attributes (which I think I agree on), we should find a way to do so also when using the by-schema validation? wdyt?

Reply:
yeah I agree. I think the kind of validation I describe is related to the business domain and should ideally live in a layer one higher than saved objects, e.g. a service exposed to other plugins or in the http controller. In such a layer you could do async validation like fetching other saved objects before accepting a create.

So an alternative could be to make sure we have schema validation in core for "root properties" like references and updated_at and then keep only validation on attributes. And then if someone needs more powerful validation the answer is "build your own http API" like we anyway already recommend.

I was thinking when all saved objects are `hidden: true` saved object validation might become obsolete, but we would always need import/export validation and this would skip any business logic in other layers. I still feel a bit uneasy about import, you wouldn't always apply the same validation as on a create since the import includes other related objects, and it becomes quite messy to enforce the same business logic on an HTTP API as on a batch of imported objects. One potential solution could be to "sign" exports with e.g. a hash of it's contents so that we know it wasn't tampered with and can safely import it, but it also removes the usefulness of this API to some extent.

Reply:
> So an alternative could be to make sure we have schema validation in core for "root properties" like references and updated_at and then keep only validation on attributes. And then if someone needs more powerful validation the answer is "build your own http API" like we anyway already recommend.

That would be aligned with our recommendations, so I'd say it make sense. 

> I was thinking when all saved objects are hidden: true saved object validation might become obsolete, but we would always need import/export validation and this would skip any business logic in other layers

When all types will have their own validation layer directly implemented in their services / endpoints, I think that import/export would be the only place remaining owned by `core` where we'll want to perform validation yes. Once this is the case, I imagine we could still allow owners to register a specific validation function / hook for this with their types, to be used specifically when validating during import or export?

> One potential solution could be to "sign" exports with e.g. a hash of it's contents so that we know it wasn't tampered with and can safely import it, but it also removes the usefulness of this API to some extent.

I think quite a few consumers are currently tampering or manually creating export files, so I'm not sure if we could introduce such thing without it being a breaking change. Plus, given that I think we'll want validation for import/export regardless, I'm not sure this would bring a significant plus value, as least for validation.

Reply:
I went ahead and added a full schema validation for the "root properties" that gets applied as well, so we are now validating the whole object, not just attributes.